langserver/handlers/execute_command_rootmodules_test.go

```diff
@@ -0,0 +1,218 @@
+package handlers
+
+import (
+	"fmt"
+	"path/filepath"
+	"testing"
+
+	"github.com/hashicorp/terraform-ls/internal/lsp"
+	"github.com/hashicorp/terraform-ls/internal/terraform/rootmodule"
+	"github.com/hashicorp/terraform-ls/langserver"
+)
+
+func TestLangServer_workspaceExecuteCommand_rootmodules_argumentError(t *testing.T) {
+	tmpDir := TempDir(t)
+	testFileURI := fmt.Sprintf("%s/main.tf", tmpDir.URI())
+	InitPluginCache(t, tmpDir.Dir())
+
+	ls := langserver.NewLangServerMock(t, NewMockSession(&MockSessionInput{
+		RootModules: map[string]*rootmodule.RootModuleMock{
+			tmpDir.Dir(): {
+				TfExecFactory: validTfMockCalls(),
+			},
+		},
+	}))
+	stop := ls.Start(t)
+	defer stop()
+
+	ls.Call(t, &langserver.CallRequest{
+		Method: "initialize",
+		ReqParams: fmt.Sprintf(`{
+	    "capabilities": {},
+	    "rootUri": %q,
+	    "processId": 12345
+	}`, tmpDir.URI())})
+	ls.Notify(t, &langserver.CallRequest{
+		Method:    "initialized",
+		ReqParams: "{}",
+	})
+	ls.Call(t, &langserver.CallRequest{
+		Method: "textDocument/didOpen",
+		ReqParams: fmt.Sprintf(`{
+		"textDocument": {
+			"version": 0,
+			"languageId": "terraform",
+			"text": "provider \"github\" {}",
+			"uri": %q
+		}
+	}`, testFileURI)})
+
+	ls.CallAndExpectError(t, &langserver.CallRequest{
+		Method: "workspace/executeCommand",
+		ReqParams: `{
+		"command": "rootmodules"
+	}`}, rootmodulesCommandURIArgNotFound.Err())
+}
+
+func TestLangServer_workspaceExecuteCommand_rootmodules_basic(t *testing.T) {
+	tmpDir := TempDir(t)
+	testFileURI := fmt.Sprintf("%s/main.tf", tmpDir.URI())
+	InitPluginCache(t, tmpDir.Dir())
+
+	ls := langserver.NewLangServerMock(t, NewMockSession(&MockSessionInput{
+		RootModules: map[string]*rootmodule.RootModuleMock{
+			tmpDir.Dir(): {
+				TfExecFactory: validTfMockCalls(),
+			},
+		},
+	}))
+	stop := ls.Start(t)
+	defer stop()
+
+	ls.Call(t, &langserver.CallRequest{
+		Method: "initialize",
+		ReqParams: fmt.Sprintf(`{
+	    "capabilities": {},
+	    "rootUri": %q,
+	    "processId": 12345
+	}`, tmpDir.URI())})
+	ls.Notify(t, &langserver.CallRequest{
+		Method:    "initialized",
+		ReqParams: "{}",
+	})
+	ls.Call(t, &langserver.CallRequest{
+		Method: "textDocument/didOpen",
+		ReqParams: fmt.Sprintf(`{
+		"textDocument": {
+			"version": 0,
+			"languageId": "terraform",
+			"text": "provider \"github\" {}",
+			"uri": %q
+		}
+	}`, testFileURI)})
+
+	ls.CallAndExpectResponse(t, &langserver.CallRequest{
+		Method: "workspace/executeCommand",
+		ReqParams: fmt.Sprintf(`{
+		"command": "rootmodules",
+		"arguments": ["uri=%s"] 
+	}`, testFileURI)}, fmt.Sprintf(`{
+		"jsonrpc": "2.0",
+		"id": 3,
+		"result": {
+			"version": 0,
+			"doneLoading": true,
+			"rootModules": [
+				{
+					"uri": %q
+				}
+			]
+		}
+	}`, tmpDir.URI()))
+}
+
+func TestLangServer_workspaceExecuteCommand_rootmodules_multiple(t *testing.T) {
```

Comment:
This test is the one failing, I imagine because mocking is used, my test scenario, where 2 rootmodules refering a common module, is not actually creating any association and terraform isn't actually reporting 2 rootmodules for the rootmodule manager? cc @radeksimko

Reply:
I don't think there's any mocking involved in the root module discovery in particular, which I believe is _actually_ the reason why this is failing. No root modules get discovered because all of these files only exist in the virtual FS.

`didOpen` doesn't create these files on the disk. Technically we would be able to do `init` for virtual files as part of tests, but we do not because users won't be able to do that anyway (because Terraform CLI has no way of reading from a virtual FS), so there is no point in testing unrealistic scenarios.

I think the best way forward here would be just creating `testdata` with regular files representing the hierarchy.